lib/semantics/expression.cc

```diff
@@ -1559,17 +1581,17 @@ static bool CheckCompatibleArguments(
   return true;
 }
 
-const Symbol *ExpressionAnalyzer::ResolveGeneric(const Symbol &symbol,
-    ActualArguments &actuals, const semantics::Scope &scope) {
+const Symbol *ExpressionAnalyzer::ResolveGeneric(
+    const Symbol &symbol, ActualArguments &actuals) {
   const Symbol *elemental{nullptr};  // matching elemental specific proc
-  const auto &details{symbol.get<semantics::GenericDetails>()};
+  const auto &details{symbol.GetUltimate().get<semantics::GenericDetails>()};
   for (const Symbol *specific : details.specificProcs()) {
     if (std::optional<characteristics::Procedure> procedure{
             characteristics::Procedure::Characterize(
                 ProcedureDesignator{*specific}, context_.intrinsics())}) {
       ActualArguments localActuals{actuals};
       auto messages{semantics::CheckExplicitInterface(
-          *procedure, localActuals, GetFoldingContext(), scope)};
+          *procedure, localActuals, GetFoldingContext(), symbol.owner())};
```

Comment:
Does this change cause `CheckExplicitInterface` to enforce modifiability checks on actual arguments from the perspective of the called procedure rather than the call site?  Will constraints on calls *from* `PURE` subprograms still work?

Reply:
I believe these still work as before. Here is a test: it gets the same errors as before (indicated in comments).
```
module m
  real, protected :: x
  real :: y
  interface s
    subroutine s1(x)
      real, intent(out) :: x
    end
  end interface
end

subroutine test1
  use m
  !ERROR: No specific procedure of generic 's' matches the actual arguments
  call s(x)
end

pure subroutine test2
  use m
  !ERROR: No specific procedure of generic 's' matches the actual arguments
  call s(y)
end
```

Reply:
These could be improved. Ideally we would resolve `s` to `s1` and then report the real errors about why the actual arguments aren't okay. That may mean we can't use `CheckExplicitInterface` in resolving generics.

Reply:
I guess it works because the specific procedure is brought into the scope of the call via USE or host association.  Would it work if the specific procedure were internal?